### ISSUE TITLE
Async all the way down... from where? Earliest point to bootstrap an async WPF application?

### ISSUE THREAD

**Issue by sam-wheat** [dotnet/wpf#6608]
What is the correct way and earliest point to bootstrap the WPF applcation shown below?  [Runnable sample project here](https://github.com/sam-wheat/WPF_AsyncExample).

To make DI and logging work I created an `IHost` after the app starts. However I feel like the correct solution to what I am asking may be to create the `IHost` first and bootstrap the app from `.RunAsync`.  I am unable to locate any documentation that illustrates how this should be done.  If it is the correct solution, why do the default templates not use it?  Do I continue to use `<UseWPF>true</UseWPF>` in the solution file?

	public partial class App : Application
	{
		protected override void OnStartup(StartupEventArgs e) => OnStartupAsync(e).Wait(); // Deadlock
		// protected override void OnStartup(StartupEventArgs e) => Task.Run(async () => await OnStartupAsync(e)).Wait();
		// protected override void OnStartup(StartupEventArgs e) => Dispatcher.InvokeAsync(async () => await OnStartupAsync(e)).Wait();
		// Need a turtle here.  

		protected async Task OnStartupAsync(StartupEventArgs e)
		{
			base.OnStartup(e);

			try
			{
				IHost host = CreateHostBuilder(e.Args).Build();
				using (ILifetimeScope scope = host.Services.GetAutofacRoot())
				{
					MainWindow mainWindow = scope.Resolve<MainWindow>();
					await mainWindow.Initialize();
					mainWindow.Show(); 
				}
			}
			catch (Exception ex)
			{
				string y = ex.ToString();
			}
		}

		public IHostBuilder CreateHostBuilder(string[] args) =>
			Host.CreateDefaultBuilder(args)
			.UseServiceProviderFactory(new AutofacServiceProviderFactory())
			.ConfigureContainer<ContainerBuilder>((hostBuilder, builder) => ConfigureContainer(hostBuilder.Configuration, builder));

		protected void ConfigureContainer(IConfiguration config, ContainerBuilder builder)
		{
			builder.RegisterType<MainWindow>().SingleInstance();
			builder.RegisterType<MainWindowViewModel>().SingleInstance();
			builder.RegisterType<UserControl1>();
			builder.RegisterType<UserControl2>();
		}
	}

There are a lot of questions related to this and a lot of documentation so I feel obligated to show my research in the links at the bottom.  Cleary's ["Turtles all the way down"](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming) non-answer nicely illustrates the abundance of literature and lack of information.  

https://stackoverflow.com/questions/12556993/significance-of-declaring-a-wpf-event-handler-as-async-in-c-sharp-5
https://weblog.west-wind.com/posts/2022/Apr/22/Async-and-Async-Void-Event-Handling-in-WPF
https://stackoverflow.com/questions/31414563/how-do-i-execute-an-asynchronous-startup-on-a-wpf-application
https://stackoverflow.com/questions/49701102/wpf-app-run-async-task-before-opening-window
Does not cover async startup:
https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.startup?view=windowsdesktop-6.0
Don't want a workaround
https://docs.microsoft.com/en-us/archive/msdn-magazine/2014/march/async-programming-patterns-for-asynchronous-mvvm-applications-data-binding
Does not cover WPF/Application:
https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/november/asynchronous-programming-async-from-the-start
https://www.jbsoftware.ca/going-async-in-uwp-and-wpf-keep-your-main-thread-free/
https://stackoverflow.com/questions/32257640/how-do-i-handle-async-operations-in-startup-configure

https://andrewlock.net/series/running-async-tasks-on-app-startup-in-asp-net-core/
https://andrewlock.net/running-async-tasks-on-app-startup-in-asp-net-core-3/

Example project is very similar to what is here
https://laurentkempe.com/2019/09/03/WPF-and-dotnet-Generic-Host-with-dotnet-Core-3-0/

Don’t block on Tasks; use async all the way down.
https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html

["Why would you want to create your WPF app from an async main anyways?"](https://github.com/dotnet/wpf/issues/997)
Because if its async all the way down its async all the way up as well.

**Comment by lindexi** [comment]
@sam-wheat Why you wait OnStartupAsync in OnStartup? The app will run and wait the message which will do not thing before you create and show the main windows.

Just call the OnStartupAsync without `Wait()`.

```csharp
protected override void OnStartup(StartupEventArgs e) => OnStartupAsync(e);
protected async void OnStartupAsync(StartupEventArgs e)
{
   ...
}
```

**Comment by sam-wheat** [comment]
Thanks for your response @lindexi.

I modified the [example code](https://github.com/sam-wheat/WPF_AsyncExample/blob/main/WPF_AsyncExample/App.xaml.cs) per your suggestion.  I also set the the shutdown mode to ensure WPF will listen for it.  I used this pattern in a dotnet Framework WPF app in the past and it worked.  IDK why it is not working here.  The code does not wait and when I try to resolve a control from the container I get an error that the container has been disposed.  

Steps to reproduce:
Clone the example repo.
Put a breakpoint on the last bracket of the try block in App.xaml.cs.
Run the project -  Breakpoint is hit indicating the method is not awaited.  Error that the container has been disposed when trying to resolve an injected control.

Expected:
mainWindow.Show() blocks and the DI container is not disposed until the app is shut down.
